### ISSUE TITLE
context.getScope() on ForStatement

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint#10246]
> I'm sorry that I don't follow issue templates.

From #10227.

```js
for (let i = 0; i < 10; ++i)
    ;
for (const key in obj)
    ;
for (const x of xs)
    ;
```

Those `for` loops (i.e. `ForStatement`, `ForInStatement`, and `ForOfStatement`) can define variable(s) in the itself scope. So `eslint-scope` makes `for` scope for those.

But, `context.getScope()`s in the following example don't return correct `for` scopes. Those return the parent scope of the correct `for` scopes. As a result, `context.getScope().variables` doesn't include iteration variables.

```js
exports.create = (context) => {
    return {
        ForStatement() {
            context.getScope()
        },
        ForInStatement() {
            context.getScope()
        },
        ForOfStatement() {
            context.getScope()
        },
    }
}
```

We should fix the `context.getScope()` behavior to return the correct scopes for `for` loops.

However, I think that we should take this change as a breaking change because this behavior has existed since ESLint borned.

**Comment by mysticatea** [comment]
**TSC Summary**: `context.getScope()` on any of `ForStatement`, `ForInStatement`, and `ForOfStatement` should return these `for` scope. Currently, those return the parent scope of these `for` scope.
**TSC Question**: Should we accept this change -- `context.getScope()` on any of `ForStatement`, `ForInStatement`, and `ForOfStatement` returns these `for` scope?

**Comment by mysticatea** [comment]
Note `context.getScope()` on `WithStatement` has a same problem as well. I'd like to include about `WithStatement` into this issue.

```js
exports.create = (context) => {
    return {
        WithStatement() {
            context.getScope() // should return a `with` scope,
                               // but it returns the parent scope of the `with` scope currently.
        },
    }
}
```

**Comment by not-an-aardvark** [comment]
This issue was accepted in [today's TSC meeting](https://github.com/eslint/tsc-meetings/blob/c271eee0c355058b90acb511a6e70a0c4520b121/notes/2018/2018-04-26.md#contextgetscope-on-forstatement).
